docs/knn/index.md

```diff
@@ -246,3 +246,98 @@ All parameters are required.
 The standard KNN query and custom scoring option perform differently. Test using a representative set of documents to see if the search results and latencies match your expectations.
 
 Custom scoring works best if the initial filter reduces the number of documents to no more than 20,000. Increasing shard count can improve latencies, but be sure to keep shard size within [the recommended guidelines](../elasticsearch/#primary-and-replica-shards).
+
+
+## Use Similarity methods in painless scripting
+
+Sometimes users would like to go beyond Elasticsearch’s built-in features for scoring and might want to customize the search scores in more complex ways.
+Elasticsearch provides script_score, the ability to provide custom scores for desired documents.
+Elasticsearch Painless, a simple, secure scripting language designed specifically for this purpose.
+
+A Painless script is evaluated within a context. Painless has a strict list of allowed methods and classes per context to ensure all Painless scripts are secure.
+Now, opendistro users can use following specialized methods within Score context to customize their scores.
+
+### Cosine Similarity
+This function calculates the measure of cosine similarity between a given query vector and document vectors.
+Optionally accepts normQueryVector, to avoid repeated calculation of normalization for query vector for every filtered documents.  
```

Comment:
Optionally accepts `normQueryVector` to avoid...

Can you be more specific about why you'd add this parameter and maybe some sample values/outcomes?